### DISCUSSION TITLE
The cost of EntityBody[+F[_]] = Stream[F, Byte]

### DISCUSSION THREAD

**Discussion by RafalSumislawski** [http4s/http4s#4814] [Ideas]
http4s uses `fs.Stream[F, Byte]` as model of HTTP Entity. It's a versatile approach. A `Stream[F, Byte]` is capable of representing which are evaluated in single piece like String or byte array (I'm going to refer to this kind of entities as `strict entities`), as well as those that are evaluated in multiple pieces (I'll call them `chunked entities`). 

While, without a doubt, the ability to handle `chunked entities` is important, majority of use-cases involve only `strict entities`. `CirceEntityEncoder.circeEntityEncoder`, `TwirlInstances.contentEncoder`, `ElemInstances.xmlEncoder`, `PlayEntityEncoder.playEntityEncoder` all produce `strict entities`. `http4s-circe` offers `CirceInstances.streamJsonArrayEncoder` which is a `chunked entity`, but it's not what majority of `http4s-circe` users actually use.

By using `Stream[F, Byte]` as the model of HTTP entities, http4s commits to paying for the overheads characteristic for `chunked entities` also when dealing with `strict entities`. These overheads come in two flavors:
- computational (and allocation) overheads of fs2
- missed optimization opportunities (it's impossible to optimize for `strict entities`, because they're indistinguishable from `chunked entities`)

In an effort to increase throughput of http4s-blaze-server I developed this "trick": https://github.com/RafalSumislawski/http4s/commit/8508b97c8942a1c5afb25588bc9f6e6ba8507c9c
I decided to open a discussion instead of a PR, because this code has some obvious shortcomings (`package fs2`, reflection, dependency on internals of fs2). Yet, what it does perfectly is illustrate how with this one trick we can increase throughput of http4s-blaze-server by 40% (more detailed data at the bottom). Knowing that we have the whole entity in one chunk, means we can do a single write operation (single system call) instead of two. Avoiding fs2 competently, frees up a lot of CPU time, and reduced GC pressure. This experiment touches only writing of responses only by blaze server, but the `EntityBody` model is shared by request and response, on server side and client size of all http4s implementations, so probably similar kind of overheads can be found in other cases.

Looking at competitors of `http4s`, we can see that they tend to solve the problem by defining a sum type with separate cases for `chunked` and `strict` entities, enabling the aforementioned :
- `akka-http` (https://github.com/akka/akka-http/blob/b7f057da0923393c290a778e5043169ee88e0bef/akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala#L344)
- `zio-http` (https://github.com/dream11/zio-http/blob/main/zio-http/src/main/scala/zhttp/http/HttpData.scala#L14)

So, this is the point were I wonder if http4s could follow in their footsteps?

Also, if you are interested in merging my "trick", please let me know. (although I doubt it :P)

#### measurements

Testing methodology was similar to that described in #4761

| parallel requests | Baseline (066fff2635f95f67a61ce290ad8523df2db5a526) |  https://github.com/RafalSumislawski/http4s/commit/8508b97c8942a1c5afb25588bc9f6e6ba8507c9c |
| ------ | ------------- | ------------- |
| 256   | 225289 req/s  | 333544 req/s |
| 512   | 276336 req/s  | 395217 req/s |
| 1024 | 278821 req/s  | 384254 req/s |
| 2048 | 278223 req/s  | 392548 req/s |
| 4096 | 237297 req/s  | 350962 req/s |

**Comment by rossabaker** [comment]
Have you seen #4509?  We're already on our way to something like this.

History: when we forked Spray, the body was either an empty entity or a wrapper around an `Array[Byte]`.  We wanted to be stream-first, and saw the strict entities as special cases of streaming, and landed on a simple model.  akka-http also evolved from Spray and developed the sum-type model.  The more I've watched it over the years, the more convinced I am that akka-http got that right and we got it wrong.

Please take a look at #4509 and we can compare and contrast.

**Comment by RafalSumislawski** [comment]
#4509 looks great :)
